### PR TITLE
[ADD][9.0] Create new module stock_picking_tracking

### PR DESCRIPTION
**Pull request by lasley**
Stock Picking Tracking
======================

This module provides an Event tracking object for Stock Pickings, which
can provide for details such as delivery tracking information.

Installation
============

* Install module

Configuration
=============

None

Usage
=====

Tracking information is located in the ``Tracking`` tab of Stock Pickings

To see a data sample:
- Inventory => [Operations] All Transfers => WH/OUT/00001 => Tracking page

Let me know if you have any suggestions!

### CODE REVIEW

**Comment by lasley** [comment]
@t3ddftw please review

**Review by tedsalmon — APPROVED** [review]
👍  Looks good to me!

**Comment by lasley** [comment]
Would a PSC mind reviewing when they get a chance please?

File: stock_picking_tracking/README.rst [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+   :alt: License: AGPL-3
+
+======================
+Stock Picking Tracking
+======================
+
+This module provides an Event tracking object for Stock Pickings, which
+can provide for details such as delivery tracking information.
+
+Installation
```
Comment (pedrobaeza) [inline stock_picking_tracking/README.rst]: Remove empty sections

File: stock_picking_tracking/models/stock_picking_tracking_group.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 LasLabs Inc.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openerp import models, fields, api
+from .stock_picking_tracking_event import OPERATION_STATES
```
Comment (pedrobaeza) [inline stock_picking_tracking/models/stock_picking_tracking_group.py]: As you want to share the same states across several objects, this is not the best strategy, as on inheritance you will lose the same states on all places. Better to use this technique: https://github.com/OCA/sale-workflow/blob/eb94d71ba9f75a0f84c9435b6dac10c010a35aa7/sale_order_type/models/sale_order_type.py#L20 (it also handles translations)

File: stock_picking_tracking/models/stock_picking_tracking_group.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            if location:
+                name += ' {location}'.format(location=location)
+            if rec_id.ref:
+                name += ' - ' + rec_id.ref
+            res.append((rec_id.id, name))
+        return res
+
+    @api.multi
+    def _process_event(self, **kwargs):
+        """ Receive & process event, handing back vals for creation
```
Comment (pedrobaeza) [inline stock_picking_tracking/models/stock_picking_tracking_group.py]: s/handing/handling

File: stock_picking_tracking/models/stock_picking_tracking_location.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    _name = 'stock.picking.tracking.location'
+    _description = 'Stock Picking Tracking Location'
+
+    event_ids = fields.One2many(
+        string='Events',
+        comodel_name='stock.picking.tracking.event',
+        inverse_name='location_id',
+    )
+    city = fields.Char(required=True)
+    zip = fields.Char(required=True)
```
Comment (pedrobaeza) [inline stock_picking_tracking/models/stock_picking_tracking_location.py]: Don't use standard `zip` string, which is use by Python.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: stock_picking_tracking/models/stock_picking_tracking_group.py (hunk 3 above)
Comment (lasley) [reply stock_picking_tracking/models/stock_picking_tracking_group.py]: Handing better described what's happening than Handling, but I think "returning" is the more appropriate term instead? I'll change to that.

Also the returns string of this docblock is wrong, I'll update that too!

**Comment by lasley** [comment]
Thanks for the review @pedrobaeza - all points fixed up

@tedsalmon - please make sure a7f2ac1 ends up in your #327 (once build passes anyways)

File: stock_picking_tracking/models/stock_picking_tracking_group.py [hunk 5]
```diff
@@ -122,13 +129,14 @@ def create_event(self, message, state, location_id,
...
 
         Returns:
-            stock.picking.tracking.event Recordset of new events
+            StockPickingTrackingEvent: Recordset of new events that were
+                created.
         """
         event_obj = self.env['stock.picking.tracking.event']
         res = event_obj.browse()
-        for rec_id in self:
+        for recorc in self:
```
Comment (pedrobaeza) [inline stock_picking_tracking/models/stock_picking_tracking_group.py]: s/recorc/record

**Comment by pedrobaeza** [comment]
Should I preserve both commits on merge?

**Comment by lasley** [comment]
Nah no worries - let's squash up. @tedsalmon can just rebase onto the merged commit instead

**Comment by lasley** [comment]
😆  sorry for my dartboard testing w/ Travis here, I admittedly didn't want to fire up my v9.

We should be 🍏  this time around 🚀

**Review by LoisRForgeFlow — APPROVED** [review]
This is a good one! :+1:

**Comment by LoisRForgeFlow** [comment]
What could be good is a menuitem in setting or wherever to access to `stock.picking.tracking.location`. Because I think is good to be able to configure some of them previously, if I'm not wrong you only can create them on the go (from the stock.picking) now, right?

**Comment by lasley** [comment]
@lreficent - Yeah the creation of tracking locations is meant to be done by a connector library, but I don't see the harm in having a menu for them. I added it in the Inventory => Configuration menu.

**Review by Cedric-Pigeon — APPROVED** [review]

**Comment by Cedric-Pigeon** [comment]
@pedrobaeza I think your changes request has been included.

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**